Implementation of coursera LMS LTI support (RA-809)

kinger68 opened the pull request:
## Summary

Added support for Coursera to launch mattermost server from their courses utilizing LTI.  Coursera does not include a user name which is necessary for RIff.  This code will generate a new unique username based on the email address specified.  In cases where conflicts may occur, a numeric delineation is used to create the unique username.

Unit tests have been added for building a user and generating a unique username.

## BREAKING CHANGES
The slug names for the personal channels (for Esme edX courses that'd be the `plg` and `capstone` channels) are now generated in a different way, which might result in a different slug name. If that were to happen learner's would be added to a _new_ private channel with the same display name as the old one.

## Ticket Link
Jira-refs: [RA-809](https://esme-learning.atlassian.net/browse/RA-809)

mlippert commented on app/lti.go:
```diff
@@ -9,16 +9,15 @@ import (
-//
-//	GetLTISettings() reads the LTI Config from Plugin Config
-//
+// GetLTISettings() reads the LTI Config from Plugin Config
```
I'd contend this change was unnecessary. We haven't settled on a style for function comments in go, and in some ways this was easier to read than other function comments that we've created. Since this function wasn't being modified in any other way maybe we should leave it along until we're ready to add function comments consistently

mlippert commented on i18n/en.json:
```diff
@@ -2050,15 +2050,15 @@
-    "id": "edx.build_user.email_missing",
+    "id": "lms.build_user.email_missing",
     "translation": "LTI-23: LTI Launch Data is missing Email."
   },
   {
-    "id": "edx.build_user.username_missing",
+    "id": "lms.build_user.username_missing",
     "translation": "LTI-24: LTI Launch Data is missing Username."
   },
   {
-    "id": "edx.build_user.lti_user_id_missing",
+    "id": "lms.build_user.lti_user_id_missing",
```
Just checking if you verified this change. I made the name change, but I hadn't really settled on how or what it should be, and so I hadn't checked that I changed all uses to match. I just remember constantly changing the names in the sources as different names seemed better.

mlippert commented on model/lti.go:
```diff
@@ -119,3 +130,38 @@ func truncateLMSChannelSlug(channelSlug string) string {
+
+	// Strip off a '+' and all characters after that from the email name for the base user name
+	username := strings.Split(emailName, "+")[0]
+	if len(username) > UserNameMaxLength {
+		username = transformLTIUsername(username)[:UserNameMaxLength]
+	} else {
+		username = transformLTIUsername(username)
+	}
+
+	return username
```
```suggestion
        username = transformLTIUsername(username)
	if len(username) > UserNameMaxLength {
		username = username[:UserNameMaxLength]
	}
	
	return username
```

mlippert commented on web/lti.go:
```diff
@@ -61,16 +62,19 @@ func loginWithLTI(c *Context, w http.ResponseWriter, r *http.Request) {
 			c.Err = err
 			return
 		}
+		launchData["lis_person_sourcedid"] = user.Username
```
It seems the normal style is to have a blank line after the end `}` of an if block. (edit: although as I look at other files it seems this is more the style in this file than everywhere, so perhaps not worth changing)
```suggestion

		launchData["lis_person_sourcedid"] = user.Username
```

mlippert commented on app/lti.go:
```diff
@@ -150,6 +149,80 @@ func (a *App) GetLTIUser(ltiUserID, email string) *model.User {
+
+	username, isUniqueUsername := lms.GetUsername(launchData)
+	if !isUniqueUsername {
+		generatedUsername, err := a.generateFreeUsername(username)
+		if err != nil {
+			return nil, err
+		}
+		username = generatedUsername
```
```suggestion
		var err
		username, err = a.generateFreeUsername(username); err != nil {
			return nil, err
		}
```

Commit:
Add coursera LMS LTI support

In the course of adding coursera these fixes were also done, the first is very likely
a possible **BREAKING CHANGE** for edX courses depending on the personal channel id's being sent!
- fix GetLMSChannelSlug to handle using id's w/ uppercase letters
- fix logging lti channel creation failures

make vet and check-style are almost clean, there is 1 remaining check-style issue
related to the HACK implemented to skip LTI validation of launchData in the cookie

```
api4/lti.go:107:6: `addLaunchDataToForm` is unused (deadcode)
func addLaunchDataToForm(ltiLaunchData map[string]string, request *http.Request) *http.Request {
     ^
```

The future plan is to remove the HACK which modifies the launch data in that cookie,
and pass the computed data (username) to the frontend in a different cookie.

kinger68 committed:
rebuilt AppIface by running Makefile target 'app-layers'

mlippert commented:
force pushed in order to squash and split out the AppIface generated files and regenerate them using `make app-layers` in a separate commit.

mlippert commented:
I just pushed some minor code review changes for @kinger68 to look over

mlippert reviewed (approved):
ok there was a flurry of additional changes, mainly to fix an issue w/ creating the private personal channels, so I'm re-approving, before the squashing happens.
